File: common/parent_pipe.py
import sys, ujson, traceback

class StdioServer(object):

    def __init__(self):
        self.pipein = sys.stdin
        self.pipeout = sys.stdout
        sys.stdout = sys.stderr   # so print works

    def run(self):
        while True:
            msg = self.rx()
            if msg is None: break

            method = msg.get('method', None)
            if method is not None:
                try:
                    methodFunc = getattr(self, 'rpc_' + method)
                    result = methodFunc(*msg.get('params'))
                    self.tx({'id': msg['id'], 'result': result, 'error': None})
                    continue
                except:
                    exctype, value, tb = sys.exc_info()
                    traceback.print_exception(exctype, value, tb, 10, sys.stderr)
                    self.tx({'id': msg['id'], 'error': str(exctype) + ': ' + str(value) })
                    continue
    def rpc_handshake(self):
        return 'handshake'

    def rx(self):
        rx_line = self.pipein.readline()
        if len(rx_line) == 0: return None # EOF
        msg = ujson.decode(rx_line)
        return msg

    def tx(self, msg):
        tx_line = ujson.encode(msg, escape_forward_slashes=False)
        self.pipeout.write(tx_line)
        self.pipeout.write('\n')
        self.pipeout.flush()

File: common/parent_pipe.js
/*
  I wish I could use cluster for this, but it uses fd 3 as a bidirectional socket
  which can't be easily forwarded with ssh.
*/
'use strict';
const _ = require('lodash');
const async = require('async');
const logio = require('../common/logio');

exports.ParentJsonPipe = ParentJsonPipe;

function ParentJsonPipe(o, handlers) {
  this.handlers = _.assign({
    rpc_handshake: (cb) => {
      cb(null, 'handshake');
    },
  }, handlers);
  this.stdin = process.stdin;
  this.stdout = process.stdout;

  let datas = [];
  this.stdin.on('data', (buf) => {
    while (buf.length) {
      let eol = buf.indexOf(10); // newline
      if (eol < 0) {
        datas.push(buf);
        return;
      } else {
        datas.push(buf.slice(0, eol));
        let rx = JSON.parse(datas.join(''));
        datas = [];
        this.handleRx(rx);
        buf = buf.slice(eol+1);
      }
    }
  });
}

ParentJsonPipe.prototype.tx = function(tx) {
  this.stdout.write(JSON.stringify(tx));
  this.stdout.write('\n');
};

ParentJsonPipe.prototype.emitInParent = function(...params) {
  this.tx({
    cmd: 'emit',
    params,
  });
};

ParentJsonPipe.prototype.handleRx = function(rx) {
  if (rx.method) {
    const cb = (err, result) => {
      if (err && err instanceof Error) err = err.message;
      this.tx({ id: rx.id, error: err, result: result });
    };
    let methodFunc = this.handlers[`rpc_${rx.method}`];
    if (!methodFunc) {
      logio.E('parent', 'No such method', rx.method);
      return cb('No such method', null);
    }
    methodFunc.call(this.handlers, ...rx.params, cb);
  }
};
